rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+
+* `nix flake check`: Do some checks on the flake (e.g. check that all
+  `packages` are really packages), then build the `checks`
+  derivations. For example, `nix flake check patchelf` fetches the
+  `patchelf` flake, evaluates it and builds it.
+
+* `nix flake clone`: Do a `git clone` to obtain the source of the
+  specified flake, e.g. `nix flake clone dwarffs` will yield a clone
+  of the `dwarffs` repository.
+
+## Offline use
+
+Since flakes replace channels, we have to make sure that they support
+offline use at least as well. Channels only require network access
+when you do `nix-channel --update`. By contrast, the `nix` command
+will periodically fetch the latest version of the global registry and
+of the top-level flake (e.g. `nix build nixpkgs:hello` may cause it to
+fetch a new version of Nixpkgs). To make sure things work offline:
+
+* It's not a failure if we can't fetch the registry.
+
+* It's not a failure if we can't fetch the latest version of a
+  flake. (For example, `fetchGit` no longer fails if we already have
+  the specified repository and `ref`.)
+
+* Fetched flakes are registered as garbage collector roots, so running
+  the garbage collector while on an airplane will not ruin your day.
+
+* There is an option `--no-net` to explicitly prevent updating the
+  registry or doing other things that need the network (such as
+  substituting). This is the default if there are no configured
+  non-loopback network interfaces.
+
+## CI integration
+
+Flakes are useful for CI systems since `flake.nix` and the lock file
+exactly specify the input repositories. For Hydra this means that you
+no longer need to configure any jobset inputs; just a top-level flake
+reference is enough. For example, the [configuration of the `patchelf`
+jobset](https://hydra.nixos.org/jobset/patchelf/master#tabs-configuration)
+just points to `github:NixOS/patchelf` and doesn't need to specify
+`nixpkgs`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Pure evaluation breaks certain workflows. In particular, it breaks the
+use of the Nixpkgs configuration file. Similarly, there are people who
+rely on `$NIX_PATH` to pass configuration data to NixOps
+configurations.
+
+# Alternatives
+[alternatives]: #alternatives
+
+For composition of multi-repository projects, the main alternative is
+to continue on with explicit `fetchGit` / `fetchTarball` calls to pull
+in other repositories. However, since there is no explicit listing of
+dependencies, this does not provide automatic updating (i.e. there is
+no equivalent of `nix flake update`).
+
+Instead of a `flake.nix`, flakes could store their metadata in a
+simpler format such as JSON or TOML. This avoids the Turing tarpit
+where getting flake metadata requires the execution of an arbitrarily
+complex, possibly non-terminating program.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* How to handle the system type? Currently `x86_64-linux` is
```

Comment:
Injecting the `localSystem` is probably a necessity to make this practical. Maybe it can be a default argument instead of being attached to a builtin. Then if the user wants to cross-build or switch the localSystem they can always pass `-A localSystem something-else` or `-A crossSystem ...`

Reply:
One possibility is to extend the flake `outputs` function with a set of arguments, e.g.
```
outputs = { inputs, args }: 
  with import inputs.nixpkgs { system = args.system; };
  { ...
  };
```
The `args.systems` argument would be provided automatically by the user interface. It would alos be part of the cache key so evaluations of the same flake for different systems wouldn't collide with each other.

The alternative is to lazily return packages for all supported platform, e.g.
```
outputs = inputs: {
  packages.hello.x86_64-linux = ...;
  packages.hello.x86_64-darwin = ...;
};
```
(This is similar to what Nixpkg's Hydra jobset does.) Then the UI could automatically select a package for the local system (e.g. `nix install flake:hello` would select `packages.hello.<my-system>`).

Reply:
One piece of positive feedback: I can use uses for both approaches but happily the `with import inputs.nixpkgs { system = ... }` approach works today on Darwin just by hardcoding `system = "x86_64-darwin";`, at least for the simple flake I'm testing with!